src/Rule/InEnum.php

```diff
@@ -0,0 +1,168 @@
+use Yiisoft\Validator\Rule\Trait\WhenTrait;
+use Yiisoft\Validator\DumpedRuleInterface;
+use Yiisoft\Validator\SkipOnEmptyInterface;
+use Yiisoft\Validator\SkipOnErrorInterface;
+use Yiisoft\Validator\WhenInterface;
+
+/**
+ * Defines validation options to check that the value is one of the values (or names) contained in an enum of the
+ * specified class.
+ * If the {@see In::$not} is set, the validation logic is inverted and the rule will ensure that the value
+ * is NOT one of them.
+ *
+ * In case of the validated value being a list, the order of values is important.
+ *
+ * Nested arrays are supported in the validated value (the order of values in
+ * lists must match, the order of keys in associative arrays is not important).
+ *
+ * @see InEnumHandler
+ *
+ * @psalm-import-type SkipOnEmptyValue from SkipOnEmptyInterface
+ * @psalm-import-type WhenType from WhenInterface
+ */
+#[Attribute(Attribute::TARGET_PROPERTY | Attribute::IS_REPEATABLE)]
+final class InEnum implements DumpedRuleInterface, SkipOnErrorInterface, WhenInterface, SkipOnEmptyInterface
+{
+    use SkipOnEmptyTrait;
+    use SkipOnErrorTrait;
+    use WhenTrait;
+
+    /**
+     * @param string $class Class of the enum to user.
+     * @param bool $useNames Whether to use names for backend enums instead of value.
+     * @param bool $strict Whether the comparison to each value in the set is strict:
+     *
+     * - Strict mode uses `===` operator meaning the type and the value must both match.
+     * - Non-strict mode uses `==` operator meaning that type juggling is performed first before the comparison. You can
+     * read more in the PHP docs:
+     *
+     * - {@link https://www.php.net/manual/en/language.operators.comparison.php}
+     * - {@link https://www.php.net/manual/en/types.comparisons.php}
+     * - {@link https://www.php.net/manual/en/language.types.type-juggling.php}
+     *
+     * Defaults to `false` meaning non-strict mode is used.
+     * @param bool $not Whether to invert the validation logic. Defaults to `false`. If set to `true`, the value must NOT
+     * be among the list of {@see $values}.
+     * @param string $message Error message when the value is not in a set of value.
+     *
+     * You may use the following placeholders in the message:
+     *
+     * - `{attribute}`: the name of the attribute.
+     * @param bool|callable|null $skipOnEmpty Whether to skip this rule if the value validated is empty.
+     * See {@see SkipOnEmptyInterface}.
+     * @param bool $skipOnError Whether to skip this rule if any of the previous rules gave an error.
+     * See {@see SkipOnErrorInterface}.
+     * @param Closure|null $when A callable to define a condition for applying the rule.
+     * See {@see WhenInterface}.
+     *
+     * @psalm-param SkipOnEmptyValue $skipOnEmpty
+     * @psalm-param WhenType $when
+     */
+    public function __construct(
+        private string $class,
+        private bool $useNames = false,
```

Comment:
Should we throw exception when non-typed enums used with disabled `useNames` option?

Reply:
No, I think it's OK to ignore this case.